osquery/tables/applications/posix/prometheus_metrics.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <sstream>
+
+#include <osquery/config.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <osquery/core/conversions.h>
+
+#include <osquery/tables/applications/posix/prometheus_metrics.h>
+#include <osquery/tables/applications/posix/prometheus_metrics_utils.h>
+
+namespace http = boost::network::http;
+
+namespace osquery {
+namespace tables {
+
+void parseScrapeResults(const std::map<std::string, retData>& scrapeResults,
+                        QueryData& rows) {
+  for (auto const& target : scrapeResults) {
+    std::stringstream ss(target.second.content);
+    std::string dest;
+    std::string ts(std::to_string(target.second.timestampMS.count()));
+
+    while (std::getline(ss, dest)) {
+      if (!dest.empty() && dest[0] != '#') {
+        std::vector<std::string> metric(osquery::split(dest, " "));
+
+        if (metric.size() > 1) {
+          Row r;
+          r[kcol_target_name] = target.first;
+          r[kcol_timestamp] = ts;
+          r[kcol_metric] = metric[0];
+          r[kcol_value] = metric[1];
+
+          rows.push_back(r);
+        }
+      }
+    }
+  }
+}
+
+void scrapeTargets(std::map<std::string, retData>& scrapeResults) {
+  http::client client(
+      http::client::options().follow_redirects(true).timeout(1));
+
+  for (auto& target : scrapeResults) {
+    try {
+      http::client::request request(target.first);
+      http::client::response response(client.get(request));
+
+      target.second.timestampMS =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              std::chrono::system_clock::now().time_since_epoch());
+      target.second.content = static_cast<std::string>(body(response));
+
+    } catch (std::exception& e) {
+      LOG(ERROR) << "failed on scrape of target " << target.first << ": "
+                 << e.what();
+    }
+  }
+}
+
+QueryData genPrometheusMetrics(QueryContext& context) {
+  QueryData result;
+  // get urls from config
+  auto parser = Config::getParser("prometheus_targets");
+  if (parser == nullptr || parser.get() == nullptr) {
+    return result;
+  }
+
+  const auto& config = parser->getData().get_child(kConfigParserRootKey);
```

Comment:
Nah, if this does somehow magically throw in the wild we don't want the daemon to come to a grinding halt. For this, use a non-throwing call with a default value like 'get_child(kConfigParserRootKey,"")' and check the output.